cpp/order/SolLiq.cc

```diff
@@ -168,65 +173,69 @@ void SolLiq::computeClustersQdot(const locality::NeighborList* nlist, const vec3
-                        Qlmjnorm = sqrt(Qlmjnorm);
-                        Qdot = Qdot / real((Qlminorm * Qlmjnorm));
-                        m_qldot_ij.push_back(Qdot); // Only i < j, other pairs not added.
-                        // Check if we're bonded via the threshold criterion
-                        if (real(Qdot) > m_Qthreshold)
-                        {
-                            // Tick up counts of number of connections these particles
-                            // have
-                            m_number_of_connections.get()[i]++;
-                            m_number_of_connections.get()[j]++;
-                        }
-                    }
-                }
-            }
-        }
-    }
+    tbb::parallel_for(tbb::blocked_range<size_t>(0, Np), [=](const tbb::blocked_range<size_t>& r) {
+
+      size_t bond(nlist->find_first_index(r.begin()));
+      for (size_t i = r.begin(); i != r.end(); i++)
+      {
+          vec3<float> p = points[i];
+
+          for (; bond < nlist->getNumBonds() && neighbor_list[2 * bond] == i; ++bond)
+          {
+              const size_t j(neighbor_list[2 * bond + 1]);
+              {
+                  if (i < j)
+                  {
+                      vec3<float> delta = m_box.wrap(p - points[j]);
+                      float rsq = dot(delta, delta);
+
+                      if (rsq < rmaxsq) // Check distance for candidate i,j
+                      {
+                          // Calc Q dotproduct.
+                          std::complex<float> Qdot(0.0, 0.0);
+                          std::complex<float> Qlminorm(0.0, 0.0); // Qlmi norm sq
+                          std::complex<float> Qlmjnorm(0.0, 0.0);
+                          for (unsigned int k = 0; k < (elements); ++k) // loop over m
+                          {
+                              Qdot += m_Qlmi_array.get()[(elements) *i + k]
+                                  * conj(m_Qlmi_array.get()[(elements) *j + k]);
+                              Qlminorm += m_Qlmi_array.get()[(elements) *i + k]
+                                  * conj(m_Qlmi_array.get()[(elements) *i + k]);
+                              Qlmjnorm += m_Qlmi_array.get()[(elements) *j + k]
+                                  * conj(m_Qlmi_array.get()[(elements) *j + k]);
+                          }
+                          Qlminorm = sqrt(Qlminorm);
+                          Qlmjnorm = sqrt(Qlmjnorm);
+                          Qdot = Qdot / real((Qlminorm * Qlmjnorm));
+                          m_qldot_ij[bond] = Qdot;
+                          // Check if we're bonded via the threshold criterion
+                          if (real(Qdot) > m_Qthreshold)
+                          {
+                              // Tick up counts of number of connections these particles
+                              // have
+                              m_number_of_connections.get()[i]++;
+                              m_number_of_connections.get()[j]++;
```

Comment:
I don't think this line is thread-safe. You could be adding to particle `j` on multiple different threads (since any `i` could have a given `j` as a neighbor). You would need to make an `enumerable_thread_specific<unsigned int *>` of size `Np`, then add to that on each thread, and then perform a reduction afterwards to get the right number of connections.

This task is much easier now than it used to be thanks to recent work by @jinsoo960. You can use the `util::ThreadStorage` class so that you don't need to know how to work with the relevant tbb class. For an example, see the RDF class. In particular, the following places are relevant: the declaration of [`m_local_bin_counts`](https://github.com/glotzerlab/freud/blob/master/cpp/density/RDF.h#L75); the way [it's initialized](https://github.com/glotzerlab/freud/blob/master/cpp/density/RDF.cc#L64); its usage [during accumulation](https://github.com/glotzerlab/freud/blob/master/cpp/density/RDF.cc#L251); [the reduction process](https://github.com/glotzerlab/freud/blob/master/cpp/density/RDF.cc#L110); and the [reset](https://github.com/glotzerlab/freud/blob/master/cpp/density/RDF.cc#L184). You should be able to copy that pattern almost exactly here to get thread-safe behavior here.

Reply:
The reduction step for the RDF is also parallelized. That's not necessary in this case, right?